### PR TITLE
KYLIN-5011 Detect and scatter skewed data in dict encoding step

### PR DESCRIPTION
**Pull request by zhengshengjun**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [x] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [x] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If this change need a document change, I will prepare another pr against the `document` branch
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at user@kylin or dev@kylin by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 1]
```diff
@@ -3125,4 +3125,32 @@ public int getRepartitionNumAfterEncode() {
     public boolean rePartitionEncodedDatasetWithRowKey() {
         return Boolean.valueOf(getOptional("kylin.engine.spark.repartition.encoded.dataset", "false"));
     }
+
+    /*
+     * Detect dataset skew in dictionary encode step.
+     * */
+    public boolean detectDataSkewInDictEncodingEnabled() {
+        return Boolean.valueOf(getOptional("detect.data.skew.in.dict.encoding", "false"));
```
Comment (hit-lacus) [inline core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:3133]: Use something like `kylin.dictionary.detect.data.skew.in.encoding`

File: core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 2]
```diff
@@ -3125,4 +3125,32 @@ public int getRepartitionNumAfterEncode() {
...
+    }
+
+    /*
+    * In some data skew cases, the repartition step during dictionary encoding will be slow.
+    * We can choose to sample from the dataset to detect skewed. This configuration is used to set the sample rate.
+    * */
+    public double sampleRateInEncodingSkewDetection() {
```
Comment (hit-lacus) [inline core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:3140]: kylin.dictionary.data.skew.detect.sample.rate

File: core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 3]
```diff
@@ -3125,4 +3125,32 @@ public int getRepartitionNumAfterEncode() {
...
+    * In KYLIN4, dictionaries are hashed into several buckets, column data are repartitioned by the same hash algorithm
+    * during encoding step too. In data skew cases, the repartition step will be very slow. Kylin will automatically
+    * sample from the source to detect skewed data and repartition these skewed data to random partitions.
+    * This configuration is used to set the skew data threshhold, valued from 0 to 1.
+    * e.g.
+    *   if you set this value to 0.05, for each value that takes up more than 5% percent of the total will be regarded
+    *   as skew data, as a result the skewed data will be no more than 20 records
+    * */
+    public double skewOccupationThreshHold() {
+        return Double.valueOf(getOptional("skew.occupation.threshhold", "0.05"));
```
Comment (hit-lacus) [inline core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:3154]: `kylin.dictionary.data.skew.key.percentage.threshhold`

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/builder/CubeTableEncoder.scala [hunk 4]
```diff
@@ -61,15 +66,59 @@ object CubeTableEncoder extends Logging {
...
           .mkString(SEPARATOR)
         val aliasName = structType.apply(columnIndex).name.concat(ENCODE_SUFFIX)
-        val encodeCol = dict_encode(col(encodeColRef).cast(StringType), lit(dictParams), lit(bucketSize).cast(StringType)).as(aliasName)
-        val columns = partitionedDs.schema.map(ty => col(ty.name)) ++ Seq(encodeCol)
+        var encodeCol = dict_encode(col(encodeColRef).cast(StringType), lit(dictParams), lit(bucketSize).cast(StringType)).as(aliasName)
+        val columns = partitionedDs.schema.map(ty => col(ty.name))
+
+        if (seg.kylinconf.detectDataSkewInDictEncodingEnabled()) {
+          //find skewed data in dict-encoding step
+          val sampleData = ds.sample(seg.kylinconf.sampleRateInEncodingSkewDetection()).cache()
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/builder/CubeTableEncoder.scala:77]: Prune unused columns before `cache` to reduce memory usage?

File: kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/dict/NBucketDictionary.java [hunk 5]
```diff
@@ -49,6 +56,15 @@
             this.absoluteDictMap = globalDictStore.getBucketDict(versions[versions.length - 1], metainfo, bucketId);
         }
         this.relativeDictMap = new Object2LongOpenHashMap<>();
+        if (!StringUtils.isEmpty(skewDictStorageFile)) {
+            FileSystem fs = FileSystem.get(new Configuration());
+            if (fs.exists(new Path(skewDictStorageFile))) {
+                Kryo kryo =new Kryo();
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/dict/NBucketDictionary.java:62]: Leave a white space under `=`.

**Comment by hit-lacus** [comment]
Would you like to enable this feature in IT, so IT can cover this feature?

File: kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/builder/CubeTableEncoder.scala (hunk 4 above)
Comment (nickSoul) [reply kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/builder/CubeTableEncoder.scala:77]: good suggestion

**Comment by zhengshengjun** [comment]
@hit-lacus fixed, thanks for you suggestions

**Pull request merged**